File: src/tree/tree.py
"""Simple Python project-to-tree structure. Supports only .py files."""

from __future__ import annotations

import ast as _ast
import os
from dataclasses import dataclass, field
from typing import Optional

PY_EXT = ".py"


class TreeError(Exception):
    pass


class PathError(TreeError):
    pass


class NodeError(TreeError):
    pass


@dataclass
class Leaf:
    """Represents a Python file."""

    name: str = ""
    path: str = ""
    syntax_tree: Optional[_ast.Module] = field(default=None, repr=False)

    def ast(self) -> None:
        """Parse the file and set syntax_tree."""
        with open(self.path, "r") as f:
            content = f.read()
        if not content:
            return
        self.syntax_tree = _ast.parse(content, filename=self.name)


@dataclass
class Node:
    """Represents a package (directory) that can contain sub-packages."""

    name: str = ""
    nodes: list[Node] = field(default_factory=list)
    leafs: list[Leaf] = field(default_factory=list)

    def print(self, _indent: int = 0) -> None:
        """Pretty-print the project structure."""
        print(self.name)
        print("|", end="")
        for leaf in self.leafs:
            print(f"\t{leaf.name}")
        for node in self.nodes:
            node.print(_indent + 1)


def get(path: str, node: Optional[Node] = None) -> Node:
    """Return all Python files in the given path organized by directory.

    Args:
        path: Relative path to scan.
        node: Optional Node to populate. A new one is created if None.

    Returns:
        Populated Node representing the directory tree.

    Raises:
        NodeError: If node is explicitly passed as a non-Node type.
        PathError: If path is empty.
        FileNotFoundError: If path does not exist.
    """
    if node is None:
        node = Node()

    if not isinstance(node, Node):
        raise NodeError("node must be a Node instance")

    if not path:
        raise PathError("empty path")

    entries = os.listdir(os.path.join(".", path))

    node.name = _current_package(path)

    if not entries:
        return node

    dirs: list[Node] = []
    leafs: list[Leaf] = []

    for entry in sorted(entries):
        full = os.path.join(path, entry)
        if os.path.isfile(os.path.join(".", full)):
            _, ext = os.path.splitext(entry)
            if ext == PY_EXT:
                leafs.append(Leaf(name=entry))
        elif os.path.isdir(os.path.join(".", full)):
            dirs.append(Node(name=entry))

    for sub_node in dirs:
        child = get(path + "/" + sub_node.name, sub_node)
        node.nodes.append(child)

    for leaf in leafs:
        leaf.path = path + "/" + leaf.name
        node.leafs.append(leaf)

    return node


def _current_package(path: str) -> str:
    i = path.find("/")
    if i == -1:
        return path
    return path[i + 1:]

File: src/main/scala/tree/Tree.scala
package tree

import java.io.File
import java.nio.file.{Files, Paths}
import scala.concurrent.{Await, Future}
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration.Duration
import scala.meta.*

val ScalaExt = ".scala"

/** Represents a package (directory) that can contain sub-packages. */
case class Node(
    var name: String = "",
    var nodes: List[Node] = Nil,
    var leafs: List[Leaf] = Nil
)

/** Represents a Scala source file. */
case class Leaf(
    name: String = "",
    path: String = "",
    var syntaxTree: Option[Source] = None
)

/** Errors used for validation. */
case class TreeError(message: String) extends Exception(message)

object Tree:

  private val errPath = "path"
  private val errNode = "node"

  /** Returns all Scala files in the given path as a tree of Nodes and Leafs.
    *
    * Equivalent to the Go `Get` function - recursively walks directories,
    * collecting `.scala` files as Leafs and subdirectories as child Nodes.
    */
  def get(path: String, node: Option[Node]): Either[TreeError, Node] =
    validate(path, node) match
      case Some(err) => Left(err)
      case None =>
        val n = node.get
        val dir = new File("./" + path)

        if !dir.exists() || !dir.isDirectory then
          Left(TreeError(s"path not found: $path"))
        else
          val files = dir.listFiles()
          if files == null then Left(TreeError(s"cannot read directory: $path"))
          else
            n.name = currentPackage(path)

            if files.isEmpty then Right(n)
            else
              val (dirs, scalaFiles) = filterScalaFilesDirs(files.toList)

              // Process subdirectories concurrently
              val futures = dirs.map { childNode =>
                Future {
                  get(path + "/" + childNode.name, Some(childNode))
                }
              }

              val results = futures.map(f => Await.result(f, Duration.Inf))

              val error = results.collectFirst { case Left(err) => err }
              error match
                case Some(err) => Left(err)
                case None =>
                  results.foreach {
                    case Right(child) => n.nodes = n.nodes :+ child
                    case _            => ()
                  }

                  // Attach leafs with their full paths
                  for leaf <- scalaFiles do
                    n.leafs = n.leafs :+ leaf.copy(path = path + "/" + leaf.name)

                  Right(n)

  private def validate(path: String, node: Option[Node]): Option[TreeError] =
    if node.isEmpty then Some(TreeError(s"node can't be nil: $errNode"))
    else if path.isEmpty then Some(TreeError(s"empty path: $errPath"))
    else None

  private def currentPackage(path: String): String =
    val i = path.indexOf('/')
    if i == -1 then path
    else path.substring(i + 1)

  private def filterScalaFilesDirs(files: List[File]): (List[Node], List[Leaf]) =
    val leafs = files.collect {
      case f if f.isFile && f.getName.endsWith(ScalaExt) =>
        Leaf(name = f.getName)
    }
    val nodes = files.collect {
      case f if f.isDirectory =>
        Node(name = f.getName)
    }
    (nodes, leafs)

end Tree

extension (leaf: Leaf)
  /** Parses the Scala source file and sets the syntaxTree field.
    * Equivalent to the Go `Ast()` method.
    */
  def ast(): Either[TreeError, Leaf] =
    val path = Paths.get(leaf.path)
    if !Files.exists(path) then
      return Left(TreeError(s"file not found: ${leaf.path}"))

    val content = new String(Files.readAllBytes(path))
    if content.isEmpty then return Right(leaf)

    content.parse[Source] match
      case Parsed.Success(tree) =>
        val updated = leaf.copy(syntaxTree = Some(tree))
        Right(updated)
      case Parsed.Error(pos, msg, _) =>
        Left(TreeError(s"parse error in ${leaf.name}: $msg"))

extension (node: Node)
  /** Pretty prints the project structure.
    * Equivalent to the Go `Print()` method.
    */
  def printTree(): Unit =
    println(node.name)
    scala.Predef.print("|")
    for l <- node.leafs do
      println("\t" + l.name)
    for child <- node.nodes do
      child.printTree()
